### PR TITLE
WIP - Added prompt user for url

### LINKED ISSUE #17047: Prompt user for URL on Ctrl/Cmd-L
When the user presses Ctrl-L (or Cmd-L on mac), the glutin port should show a prompt and ask for a new URL.

We should use tinyfiledialogs-rs. Here is an example of how to use a prompt: https://github.com/jdm/tinyfiledialogs-rs/blob/4bf666ab49bf9aafcbbc34b6d4383db00d737654/examples/main.rs#L10

The reload (Ctrl-R) keybinding can be used as an example: https://github.com/servo/servo/blob/764da7ba39c0ccbaf807c6131418761f70584322/ports/glutin/window.rs#L1266

We should add a rule for Ctrl-L, and if `shell.builtin-key-shortcuts.enabled` is true, use tinyfiledialogs to ask for a new url (and make the default value of the dialog the current url).

Once this is done, the url can be sanitize (can be done in a follow-up).

And then tell Servo to load this url with `WindowEvent::LoadUrl(input)`.

### PR DESCRIPTION
**Pull request by jklepatch**
This is not ready yet, I would just like to have some feedback.

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17047 
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17062)

### CODE REVIEW

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -1268,12 +1269,33 @@ impl WindowMethods for Window {
                     self.event_queue.borrow_mut().push(WindowEvent::Reload);
                 }
             }
+            //On windows 8.1 when pressing ctrl + anything, ch = Some('\u{unicode_value}')
```
Comment (paulrouget) [inline ports/glutin/window.rs]: We need to investigate that before merging this PR.

**Review by paulrouget — COMMENTED** [review]
It's weird. The changes in Cargo.lock make it impossible for me to compile.

```
./mach build -d --verbose
cargo build -v
error: failed to parse lock file at: /Users/paul/git/servo/Cargo.lock

Caused by:
  package `gcc 0.3.45 (registry+https://github.com/rust-lang/crates.io-index)` is specified as a dependency, but is missing from the package list
```

I'll see later what's going on.

File: ports/glutin/window.rs [hunk 2]
```diff
@@ -1402,6 +1424,10 @@ fn is_printable(key_code: VirtualKeyCode) -> bool {
     }
 }
 
+fn sanitize_url(url: String) -> Result<ServoUrl, ()> {
+    Ok(ServoUrl::parse(url.trim()).unwrap())
```
Comment (paulrouget) [inline ports/glutin/window.rs]: FYI, unwrap() on a failing Result will panic. See https://is.gd/RAdLog

File: ports/glutin/window.rs [hunk 3]
```diff
@@ -1268,12 +1269,33 @@ impl WindowMethods for Window {
...
                 }
             }
+            //On windows 8.1 when pressing ctrl + anything, ch = Some('\u{unicode_value}')
+            //so matching ch with Some('l') doesn't work here
+            //Instead we match with what's inside the key variable
+            (CMD_OR_CONTROL, _, Key::L) => {
+                if let Some(true) = PREFS.get("shell.builtin-key-shortcuts.enabled").as_boolean() {
+                    let user_input: String;
+                    //TODO: 
+                    // 1) make it work when clicking on "validate" button - Currently only works when pressing enter
```
Comment (paulrouget) [inline ports/glutin/window.rs]: Maybe it has to do with the fact we're blocking the main thread.
On Mac, the buttons work, but we get the beatchball.

Maybe this code should run in a different thread.

File: ports/glutin/window.rs [hunk 4]
```diff
@@ -1402,6 +1419,10 @@ fn is_printable(key_code: VirtualKeyCode) -> bool {
     }
 }
 
+fn sanitize_and_parse_url(url: String) -> Result<ServoUrl, ()> {
```
Comment (paulrouget) [inline ports/glutin/window.rs]: Actually, we might want to use `ParseError` for now instead of an empty error.

File: ports/glutin/window.rs [hunk 5]
```diff
@@ -1268,12 +1269,28 @@ impl WindowMethods for Window {
...
+            //On windows 8.1 when pressing ctrl + anything, ch = Some('\u{unicode_value}')
+            //so matching ch with Some('l') doesn't work here
+            //Instead we match with what's inside the key variable
+            (CMD_OR_CONTROL, _, Key::L) => {
+                if let Some(true) = PREFS.get("shell.builtin-key-shortcuts.enabled").as_boolean() {
+                    //TODO: 
+                    // 1) make it work when clicking on "validate" button - Currently only works when pressing enter
+                    // 2) Erratic behavior - sometime the dialog reloads empty after pressing enter
+                    // (windows 8.1)
+                    if let Some(user_input) = tinyfiledialogs::input_box("Enter a URL", "URL:", "")  {
```
Comment (paulrouget) [inline ports/glutin/window.rs]: The default value should be the current url, not `""`.

I'm wondering if we have any locales support. I'm not sure we want to keep english sentences in the code.

**Comment by paulrouget** [comment]
@jdm is tinyfiledialogs thread safe? It appears to work well here (osx) if I open a prompt from a different thread. But I'm not sure it's the right thing to do.

**Comment by jdm** [comment]
I do not believe the library does anything special to ensure that it works from non-main threads. I suspect that at least one platform will not work as intended if used that way.

**Comment by jklepatch** [comment]
Currently, in `handle_key()` (window.rs) I am matching `ctrl + L` on `Key` instead of `ch`. This is a temporary solution. Once this related issue [PR](https://github.com/servo/servo/issues/17146) is solved, I can match on `ch` and the PR will be ready.

**Comment by jdm** [comment]
Closing due to inactivity.

**Pull request closed without merging**